yoda/types.h

```diff
@@ -104,6 +123,112 @@ void SetKey(T_ENTRY& entry, ENTRY_KEY_TYPE<T_ENTRY> key) {
   KEY_ACCESSOR<T_ENTRY>::SetKey(entry, key);
 }
 
+// Matrix row and column type extractors, getters and setters.
+// Support two access methods:
+// - `.row / .col` data members,
+// - `.row() / set_row() / .col() / .set_col()` methods.
+template <typename T_ENTRY>
+constexpr bool HasRowFunction(char) {
+  return false;
+}
+
+template <typename T_ENTRY>
+constexpr auto HasRowFunction(int) -> decltype(std::declval<T_ENTRY>().row(), bool()) {
+  return true;
+}
+
+template <typename T_ENTRY, bool HAS_ROW_FUNCTION>
+struct ROW_ACCESSOR_IMPL {};
+
+template <typename T_ENTRY>
+struct ROW_ACCESSOR_IMPL<T_ENTRY, false> {
+  typedef decltype(std::declval<T_ENTRY>().row) T_ROW;
+  static typename std::conditional<std::is_pod<T_ROW>::value, T_ROW, const T_ROW&>::type GetRow(
+      const T_ENTRY& entry) {
+    return entry.row;
+  }
+  static void SetRow(T_ENTRY& entry, T_ROW row) { entry.row = row; }
+};
+
+template <typename T_ENTRY>
+struct ROW_ACCESSOR_IMPL<T_ENTRY, true> {
+  typedef decltype(std::declval<T_ENTRY>().row()) T_ROW;
+  static typename std::conditional<std::is_pod<T_ROW>::value, T_ROW, const T_ROW&>::type GetRow(
+      const T_ENTRY& entry) {
+    return entry.row();
+  }
+  static void SetRow(T_ENTRY& entry, T_ROW row) { entry.set_row(row); }
+};
+
+template <typename T_ENTRY>
+using ROW_ACCESSOR = ROW_ACCESSOR_IMPL<T_ENTRY, HasRowFunction<T_ENTRY>(0)>;
+
+template <typename T_ENTRY>
+typename ROW_ACCESSOR<T_ENTRY>::T_ROW GetRow(const T_ENTRY& entry) {
+  return ROW_ACCESSOR<T_ENTRY>::GetRow(entry);
+}
+
+template <typename T_ENTRY>
+using ENTRY_ROW_TYPE =
+    typename std::remove_cv<typename std::remove_reference<typename ROW_ACCESSOR<T_ENTRY>::T_ROW>::type>::type;
```

Comment:
Note to self: Make use of the shortened versions of this, which I am just about to push to `Bricks`!